Rewrite mountinfo handling logic.

Omar007 opened the pull request:
Signed-off-by: Omar Pakker <Omar007@users.noreply.github.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
This PR fixes the handling of mountpoints across mount namespaces. The current implementation assumes that paths on a filesystem are hard-coupled to their mount paths but this is not the case, especially not when crossing mount namespace boundaries. 
As a result, the current implementation had received mount point based work-arounds but still breaks as soon as the the mount point is not an exact match with the path within the filesystem. 
This rewrites the logic to use the filesystem paths instead of the mount paths and then uses that information to resolve the actual mount path afterwards. 

In summary the logic is based on the following information.
1) The device will be the same in any namespace so we start there and match the device first.
2) Since a device can be mounted on multiple locations, it is possible this will return multiple possible entries. Each of these mounts a different path of the filesystem somewhere but the path must be equal to or at a higher level. We can eliminate the ones that aren't located on the same filesystem path.
3) Since it's still possible for a given path on a filesystem to be mounted at several locations, attempt to use the longest/deepest path match we can find for the filesystem path.

Aside from this, the mount namespaces documentation (`man 7 mount_namespaces`) also specifies 3 types of mounts; private, shared and slave. When shared or slave, the mountinfo entry also provides information on the peer group they are in. It might be possible to use this information to further constrain the possible valid mountinfo entries but I don't think we need this as long as we use the device based information as our base and not the system's mount paths.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4590 #3119

**Special notes for your reviewer**:
First time using go so if there are any language specific notations I can use to make this leaner, please let me know.  

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Fixes handling of /proc/<pid>/mountpoint by working on the device information instead of mount information
```

vasiliy-ul commented on pkg/virt-handler/isolation/isolation.go:
```diff
@@ -354,37 +354,35 @@ func (r *realIsolationResult) IsBlockDevice(path string) (bool, error) {
-			}
+		if record[2] != mountInfo.DeviceContainingFilesystem ||
+			!strings.HasPrefix(mountInfo.PathOnFilesystem, record[3]) ||
+			(parentMountInfo != nil && len(record[3]) <= len(parentMountInfo.PathOnFilesystem)) {
```
I guess `parentMountInfo != nil` will always evaluate to `false` here since it is assigned bellow before breaking the iteration.

Omar007 replied:
What do you mean? If the device didn't match or the path on the filesystem on the device doesn't match, it will stay `nil` until both the device and path have been found.  
From that point on, if we find other entries that match the device an path on the filesystem we should only replace `parentMountInfo` if we found a longer match.

You are however correct in that it breaks the iteration after the first find/match. I thought the function was akin to a filter predicate but it aborts parsing the rest of the entries/records when `true` is returned. I should drop that `return true` otherwise if there is a longer match available it would never be found.

Omar007 replied:
Updated to always return false since I want to match/check all records, if statement is the same as before but inverted to not duplicate the return statement.

davidvossel reviewed (commented):
/ok-to-test

@Omar007 We have a unit test suite for parsing mountinfo examples. The test itself can be found in ```pkg/virt-handler/isolation/isolation_test.go``` and the test data sets are stored in ```pkg/virt-handler/isolation/testdata/mountinfo/```

One thing that would help ensure we both fully understand how your fix works and don't break it in the future is to add a test case there.

Omar007 commented:
> One thing that would help ensure we both fully understand how your fix works and don't break it in the future is to add a test case there.

It passed the existing ones locally on my end. Sadly we already know the test cases are indeed not sufficient as they passed previously as well but could easily break in real world usage.  
I'll see if I can add an extra dataset based on having a `/root` prefix as mentioned in #4590 (I could just mutate the existing dataset for that) and see if I can simulate/create a new dataset using chroot or something with the parent having a mount at 2 locations.  
I don't think I can really fabricate/create one that covers all real-world possible combinations though. I don't suppose there is a mountinfo test data toolset/generator in existence somewhere haha.

Omar007 commented:
I went through the test file and don't like what I'm seeing so I'll have to do a little bit of refactoring on the isolation package.  
Additionally, I saw there is a library to handle the mountinfo logic so I'll be using that as well. It seems it is also already in the Go dependencies/libraries file thing (go.sum) of the project.

I'll push the changes later tonight once I've made them.

Omar007 commented:
I think that should cover it.  

In hindsight I probably didn't need to have split the detector and the isolator types and related functions into their own files but while I was trying to figure out Go and how everything worked together it was a lot easier to just separate them. That way I didn't need to see it at all while working on this.

davidvossel commented:
/retest

davidvossel commented:
@Omar007 there are a significant number of functional tests failing on this PR that should be investigated. 

Also, coupling your logic change with a significant refactor is going to make this PR more difficult to get in.  I'm not saying that refactors are not welcome, it just complicates things when refactors are also coupled with logical changes as well. 

I think you'll have a better shot at getting this merged quickly if you split this into two PRs with a very narrow scope. One PR just to handle the bug fix, and one to handle a refactor with no expected logic changes.

Omar007 commented:
> there are a significant number of functional tests failing on this PR that should be investigated.

I noticed that earlier yea. I had only focussed on this specific area. This indicates there are some seriously important checks missing here I think. I'll see if I can trace back the reason for the other failures to this and what check would be relevant here to cover that.

> Also, coupling your logic change with a significant refactor is going to make this PR more difficult to get in. I'm not saying that refactors are not welcome, it just complicates things when refactors are also coupled with logical changes as well.

The problem was that I couldn't even get a good test in at all without making those changes. The way it was set up made it impossible for me to do so without making a few other changes.  
I could maybe do it the other way around though; the refactor first, then the fix/additions.

> I think you'll have a better shot at getting this merged quickly if you split this into two PRs with a very narrow scope. One PR just to handle the bug fix, and one to handle a refactor with no expected logic changes.

As I mentioned I can probably undo the separation of the detector types and functions but that is also the set that wasn't actually touched at all so that would not actually make a net change anywhere...

Omar007 commented:
Looks like that might actually be a very easy fix. Now that I changed stuff around it was quite easy to trace this back to the relevant location so writing a test as we speak. Will push something in a bit.

davidvossel commented:
/retest

davidvossel commented:
@Omar007, what is left here? I still see quite a bit of change has occurred here for what seemed like it should be a simple fix.

Omar007 commented:
Personally I've got nothing else to add or change at this point. Unless you want me to do one of the following:
1. attempt to split things out into a separate refactor change with no changes or fixes first and a fix+tests implementation second, which would be hard and honestly not worth the time or effort; the refactor that allowed me to fix things required me to touch broken code that had no real tests to verify if the changes would have kept the logic intact (bug-for-bug refactor or did it get worse?). Thus you'd more than likely want the second change anyway before merging anything otherwise you'd have nothing other than the existing functional tests to hopefully at least tell you that functionality as a whole wasn't impacted.
2. not use that library at all within this package and implement the library's mount(info) handling myself. This would strip the `vendor/` additions from this diff/MR but would basically require me to duplicate a set of what's in that library.
3. move the unchanged PodIsolationDetector types, functions and tests back into the isolator file instead of splitting it out to the separate detector files. This should shrink the diff some. This was mostly to help me get a feeling for Go and splitting related types+functions out into their own file for better overview and mental clearing but that is personal preference I suppose and not strictly necessary for this change (or any, as technically speaking everything here (any Go code) could just be put into a single file if you want)

I'd say this should be it.

davidvossel commented:
> attempt to split things out into a separate refactor change with no changes or fixes first and a fix+tests implementation second, which would be hard and honestly not worth the time or effort;

For me to review this, i need to clearly see what the logical changes are and for that to be separated from the refactor. The refactor and fixes can occur in the same PR, i just need to be able to clearly see what has occurred in separate commits.

Omar007 commented:
How is this? I split the detector split that didn't change any code out into its own/separate commit, followed by the mount logic rewrite of the isolation types, followed by the used library, followed by the missing test case we discovered and lastly a few bonus tests I added later.

Omar007 commented:
/retest

davidvossel commented:
/retest

davidvossel reviewed (approved):
/lgtm
/approve

Thanks @Omar007 for the contribution. Sorry for the delayed review

Omar007 commented:
> Thanks @Omar007 for the contribution. Sorry for the delayed review

Np, it ended up as quite a big one after all. Thanks for the review! :+1:
